tests/integration-test/server/server.js

```diff
@@ -0,0 +1,446 @@
+    return {
+      code: 200,
+      body: {},
+    };
+  }
+
+  function readRequestBody(req) {
+    debug('Reading request body.');
+    return new Promise(function(resolve, reject) {
+      req.setEncoding('utf8');
+      var body = '';
+      req.on('data', (data) => {
+        body += data;
+      });
+      req.on('end', () => {
+        debug('Raw body data', body);
+        if (req.headers['content-type'] === 'application/json') {
+          return resolve(JSON.parse(body));
+        } else { // assume application/x-www-form-urlencoded
+          return parseUrlEncodedBody(body)
+            .then(resolve);
+        }
+      });
+    });
+  }
+
+  function existsAndNumber(t) {
+    return t !== undefined && t !== null && typeof t === 'number';
+  }
+
+  function checkQuery(req, res, field) {
+    return new Promise(function(resolve, reject) {
+      const u = url.parse(req.url, true);
+      if (!u.query[field]) {
+        return resolve({
+          code: 404,
+          mesage: 'Not found. Invalid signed url format.',
+        });
+      }
+      return resolve(u);
+    });
+  }
+
+  function createAndPipeReadStream(url, res) {
+    return new Promise(function(resolve, reject) {
+      const rs = fs.createReadStream(url.query.path);
+      rs.on('error', (err) => {
+        return resolve({
+          code: 400,
+          body: err,
+        });
+      });
+      const filename = getQueryFilename(url.query.path);
+      debug('The proposed filename for header:', filename);
+      res.setHeader('Content-Disposition', `attachment; filename="${filename}"`);
+      res.on('finish', () => {
+        debug('The response stream completed.');
+        debug('The response object', res);
+        debug('The response headers', res.getHeaders());
+      });
+      rs.pipe(res);
+      return resolve({code: 200});
+    });
+  }
+
+  function createAndPipeWriteStream(url, req, res) {
+    return new Promise(function(resolve, reject) {
+      const ws = fs.createWriteStream(url.query.path);
+      ws.on('error', (err) => {
+        return resolve({
+          code: 400,
+          body: err,
+        });
+      });
+      ws.on('finish', () => {
+        return resolve({
+          code: 201,
+          body: {},
+        });
+      });
+      req.pipe(ws);
+    });
+  }
+
+  function parseUrlEncodedBody(raw) {
+    return new Promise(function(resolve, reject) {
+      var bodyShell = {};
+      const fields = raw.split('&');
+      debug('Split body fields', fields);
+      for (const field of fields) {
+        debug('Next field', field);
+        const parts = field.split('=');
+        bodyShell[parts[0]] = parts[1];
+      }
+      debug('The constructed body', bodyShell);
+      return resolve(bodyShell);
+    });
+  }
+
+  function generateOutputReport() {
+    return new Promise(function(resolve, reject) {
+      console.log('\n----- Test report -----\n');
+      reportAssertion('Check for task JSON retrieval.',
```

Comment:
Readability suggestions:
- use `[ ] check failed` and `[x] check passed` to provide some eye-catching visual cue if the test passed
- report `X/Y checks passed` at the end of the report as another visual cue. If all tests passed, make sure to clearly indicate it. The passing outcome is more important than the details above